Firestore/src/FirestoreSessionHandler.php

```diff
@@ -0,0 +1,404 @@
+ * `session_write_close()` explicitly and then handle `E_USER_WARNING`
+ * properly. See the second example for a demonstration.
+ *
+ * Example:
+ * ```
+ * use Google\Cloud\Firestore\FirestoreClient;
+ *
+ * $firestore = new FirestoreClient();
+ *
+ * $handler = $firestore->sessionHandler();
+ *
+ * session_set_save_handler($handler, true);
+ * session_save_path('sessions');
+ * session_start();
+ *
+ * // Then write and read the $_SESSION array.
+ * $_SESSION['name'] = 'Bob';
+ * echo $_SESSION['name'];
+ * ```
+ *
+ * ```
+ * // Session handler with error handling:
+ * use Google\Cloud\Firestore\FirestoreClient;
+ *
+ * $firestore = new FirestoreClient();
+ *
+ * $handler = $firestore->sessionHandler();
+ * session_set_save_handler($handler, true);
+ * session_save_path('sessions');
+ * session_start();
+ *
+ * // Then read and write the $_SESSION array.
+ * $_SESSION['name'] = 'Bob';
+ *
+ * function handle_session_error($errNo, $errStr, $errFile, $errLine) {
+ *     // We throw an exception here, but you can do whatever you need.
+ *     throw new RuntimeException(
+ *         "$errStr in $errFile on line $errLine",
+ *         $errNo
+ *     );
+ * }
+ * set_error_handler('handle_session_error', E_USER_WARNING);
+ *
+ * // If `write` fails for any reason, an exception will be thrown.
+ * session_write_close();
+ * restore_error_handler();
+ *
+ * // You can still read the $_SESSION array after closing the session.
+ * echo $_SESSION['name'];
+ * ```
+ *
+ * @see http://php.net/manual/en/class.sessionhandlerinterface.php SessionHandlerInterface
+ */
+class FirestoreSessionHandler implements SessionHandlerInterface
+{
+    use SnapshotTrait;
+
+    /**
+     * @var ConnectionInterface
+     */
+    private $connection;
+    /**
+     * @var ValueMapper
+     */
+    private $valueMapper;
+    /**
+     * @var string
+     */
+    private $projectId;
+    /**
+     * @var string
+     */
+    private $database;
+    /**
+     * @var array
+     */
+    private $options;
+    /**
+     * @var string
+     */
+    private $savePath;
+    /**
+     * @var string
+     */
+    private $sessionName;
+    /**
+     * @var Transaction
+     */
+    private $transaction;
+
+    /**
+     * Create a custom session handler backed by Cloud Firestore.
+     *
+     * @param ConnectionInterface $connection A Connection to Cloud Firestore.
+     * @param ValueMapper $valueMapper A Firestore Value Mapper.
+     * @param string $projectId The current project id.
+     * @param string $database The database id.
+     * @param array $options [optional]
+     */
+    public function __construct(
+        ConnectionInterface $connection,
+        ValueMapper $valueMapper,
+        $projectId,
+        $database,
+        array $options = []
+    ) {
+        $this->connection = $connection;
+        $this->valueMapper = $valueMapper;
+        $this->projectId = $projectId;
+        $this->database = $database;
+        $this->options = $options + [
```

Comment:
for reference on later comments:

* entries corresponding to service calls: `begin`, `commit`, `rollback`, `query`, `read`.
* entries which do not cause service calls: `delete`.

We should be able to remove `delete` and add `read`.

Reply:
I don't understand why, but I am okay with it

Reply:
Is it because any option that would be passed in `delete` can be passed in `commit`?

Reply:
Calls to `Transaction::delete()` (and other mutating methods on Transactions) do not immediately trigger service calls. It enqueues writes which are sent by calling `commit()`. The `$options` value on those methods allow for modifying the mutation (doing things like adding preconditions, etc), which do not apply in this context. So we only need to allow users to provide options for things that actually make service calls.